src/token.rs

```diff
@@ -662,80 +886,174 @@ impl Token {
-                    self.user_login.set_pin(&self.info, pin)
-                } else {
-                    if old.is_none() {
-                        return CKR_PIN_INCORRECT;
-                    }
-                    self.user_login.change_pin(&self.info, pin, old.unwrap())
-                };
-                if ret != CKR_OK {
-                    return ret;
-                }
-                /* update pin in storage */
-                match self.store_pin_object(
-                    USER_PIN_UID.to_string(),
-                    "User PIN".to_string(),
-                    pin.clone(),
-                ) {
-                    Ok(()) => (),
-                    Err(_) => return CKR_GENERAL_ERROR,
-                }
+    fn clear_private_session_objects(&mut self) {
+        let mut priv_handles = Vec::<CK_OBJECT_HANDLE>::new();
+        for (handle, obj) in self.session_objects.iter() {
+            if obj.is_private() {
+                priv_handles.push(*handle);
             }
-            CKU_SO => {
-                if old.is_none() {
-                    return CKR_PIN_INCORRECT;
-                }
-                let ret =
-                    self.so_login.change_pin(&self.info, pin, old.unwrap());
-                if ret != CKR_OK {
-                    return ret;
-                }
-                /* update pin in storage */
-                match self.store_pin_object(
-                    SO_PIN_UID.to_string(),
-                    "SO PIN".to_string(),
-                    pin.clone(),
-                ) {
-                    Ok(()) => (),
-                    Err(_) => return CKR_GENERAL_ERROR,
-                }
+        }
+
+        /* remove all private session objects */
+        for handle in priv_handles {
+            let _ = self.session_objects.remove(&handle);
+            self.handles.remove(handle);
+        }
+    }
+
+    fn clear_session_objects(&mut self, session: CK_SESSION_HANDLE) {
+        let mut handles: Vec<CK_OBJECT_HANDLE> = Vec::new();
+        for (_, obj) in self.session_objects.iter() {
+            if obj.get_session() == session {
+                handles.push(obj.get_handle());
             }
-            _ => return CKR_GENERAL_ERROR,
         }
 
-        /* If we set a PIN it means we switched to require Logins */
-        self.info.flags |= CKF_LOGIN_REQUIRED;
+        for h in handles {
+            let _ = self.session_objects.remove(&h);
+            self.handles.remove(h);
+        }
+    }
 
-        self.update_token_pin_flags();
+    fn encrypt_value(&self, uid: &String, val: &Vec<u8>) -> KResult<Vec<u8>> {
+        if let Some(ref kek) = self.kek {
+            let aad = uid.as_bytes();
+            let tag_len = 8;
+            let mut iv = [0u8; 12];
+            get_random_data(&mut iv)?;
+            let mut params = CK_GCM_PARAMS {
+                pIv: iv.as_ptr() as *mut CK_BYTE,
+                ulIvLen: 12,
+                ulIvBits: 96,
```

Comment:
nit: would be good to keep the 12 in sync here by using either local `iv_len` variable or even constant to share it with the deciphering code.

The `ulIvBits` is `8*iv_len` so it could also be modified to avoid another magic number. Or it can be even skipped as this is not used by the current implementation.

Reply:
going to create a helper function for CK_GCM_PARAMS for enc/dec as well, and will handle it there